protocols/autonatv2/src/request_response.rs

```diff
@@ -0,0 +1,324 @@
+use rand::Rng;
+
+use crate::generated::structs as proto;
+
+const REQUEST_MAX_SIZE: usize = 4104;
+pub(super) const DATA_LEN_LOWER_BOUND: usize = 30_000u32 as usize;
+pub(super) const DATA_LEN_UPPER_BOUND: usize = 100_000u32 as usize;
+pub(super) const DATA_FIELD_LEN_UPPER_BOUND: usize = 4096;
+
+macro_rules! new_io_invalid_data_err {
+    ($msg:expr) => {
+        io::Error::new(io::ErrorKind::InvalidData, $msg)
+    };
+}
+
+macro_rules! check_existence {
+    ($field:ident) => {
+        $field.ok_or_else(|| new_io_invalid_data_err!(concat!(stringify!($field), " is missing")))
+    };
+}
+
+macro_rules! read_from {
+    () => {
+        pub(crate) async fn read_from<R>(mut reader: R) -> io::Result<Self>
+        where
+            R: AsyncRead + Unpin,
+        {
+            let bytes = read_length_prefixed(&mut reader, 1024).await?;
+            Self::from_bytes(&bytes)
+        }
+    };
+}
+
+macro_rules! write_into {
+    () => {
+        pub(crate) async fn write_into<W>(self, mut writer: W) -> io::Result<()>
+        where
+            W: AsyncWrite + Unpin,
+        {
+            let bytes = self.into_bytes();
+            write_length_prefixed(&mut writer, bytes).await
+        }
+    };
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum Request {
+    Dial(DialRequest),
+    Data(DialDataResponse),
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct DialRequest {
+    pub(crate) addrs: Vec<Multiaddr>,
+    pub(crate) nonce: u64,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct DialDataResponse {
+    pub(crate) data_count: usize,
+}
+
+impl Request {
+    read_from!();
+
+    fn from_bytes(bytes: &[u8]) -> io::Result<Self> {
+        let mut reader = BytesReader::from_bytes(bytes);
+        let msg = proto::Message::from_reader(&mut reader, bytes)
+            .map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))?;
+        match msg.msg {
+            proto::mod_Message::OneOfmsg::dialRequest(proto::DialRequest { addrs, nonce }) => {
+                let addrs: Vec<Multiaddr> = addrs
+                    .into_iter()
+                    .map(|e| e.to_vec())
+                    .map(|e| {
+                        Multiaddr::try_from(e).map_err(|err| {
+                            new_io_invalid_data_err!(format!("invalid multiaddr: {}", err))
+                        })
+                    })
+                    .collect::<Result<Vec<_>, io::Error>>()?;
+                let nonce = check_existence!(nonce)?;
+                Ok(Self::Dial(DialRequest { addrs, nonce }))
+            }
+            proto::mod_Message::OneOfmsg::dialDataResponse(proto::DialDataResponse { data }) => {
+                let data_count = check_existence!(data)?.len();
+                Ok(Self::Data(DialDataResponse { data_count }))
+            }
+            _ => Err(new_io_invalid_data_err!(
+                "invalid message type, expected dialRequest or dialDataResponse"
+            )),
+        }
+    }
```

Comment:
Please use `quick-protobuf-codec` for this stuff.

Reply:
Changed that.